freeqdsk/aeqdsk.py

```diff
@@ -335,27 +706,268 @@ def read(fh):
-    for key, default, doc in fields:
-        if key is None:
-            continue  # skip
+    # Read first block of data
+    general_block_1_values = read_array(len(_general_block_1), fh, data_fmt)
+    for field, value in zip(_general_block_1, general_block_1_values):
+        data[field.name] = value
 
-        if callable(default):
-            default = default(data)
+    # Read laser bits
+    for field in _laser_block:
+        values = read_array(data[field.has_length], fh, data_fmt)
+        data[field.name] = values
 
-        if isinstance(default, list):
-            # Read a list the same length as the default
-            data[key] = [next(values) for elt in default]
-        else:
-            value = next(values)
-            if isinstance(default, int) and not isinstance(value, int):
-                # Expecting an integer, but didn't get one
-                warnings.warn("Expecting an integer for '" + key + "' in aeqdsk file")
-                break
-            data[key] = value
+    # Read next block of data
+    general_block_2_values = read_array(len(_general_block_2), fh, data_fmt)
+    for field, value in zip(_general_block_2, general_block_2_values):
+        data[field.name] = value
+
+    # Try reading first line of extended section. If this fails, raise a warning and
+    # return data without extended portion
+    try:
+        extended_sizes = read_array(len(_extended_sizes), fh, extended_sizes_fmt)
+    except Exception:
+        warnings.warn("Failed to read A-EQDSK extended section, assuming old file type")
+        return data
+
+    for field, value in zip(_extended_sizes, extended_sizes):
+        data[field.name] = value
+
+    # The first two arrays are joined together
+    joined_len = sum(data[field.has_length] for field in _extended_arrays_1)
+    joined = read_array(joined_len, fh, data_fmt)
+    data[_extended_arrays_1[0].name] = joined[: data[_extended_arrays_1[0].has_length]]
+    data[_extended_arrays_1[1].name] = joined[data[_extended_arrays_1[0].has_length] :]
+
+    # The next two are normal
+    for field in _extended_arrays_2:
+        values = read_array(data[field.has_length], fh, data_fmt)
+        data[field.name] = values
+
+    # Read in another general data block
+    extended_values = read_array("all", fh, data_fmt)
+    if len(extended_values) > len(_extended_general):
+        warnings.warn(
+            "Encountered variables at the end of an A-EQDSK file that are not "
+            "recognised by FreeQDSK. Please consider contributing to the project "
+            "and letting us know what these are!"
+        )
+    # Zip only includes elements up to the shortest iterable provided. data will not
+    # include elments that FreeQDSK doesn't recognise, nor will it store default values
+    # for elements not in the file.
+    for field, value in zip(_extended_general, extended_values):
+        data[field.name] = value
 
     return data
+
+
+# Dynamic docstring generation
```

Comment:
This is very fancy! It's probably worth putting a short note as to _why_ the docstring is dynamically generated